Stefanozampini/add petsc cuda support

stefanozampini opened the pull request:
This adds preliminary support for using PETSc accelerated linear algebra from MFEM

Currently works for MFEM_USE_CUDA if PETSc has been configured with cuda support.

Can be extended to ceed/gpu

examples/petsc/ex1p.cpp and examples/petsc/ex9p.cpp have been modified to work with --device cuda

jandrej commented:
I'm going to add the necessary cmake changes when this gets approved.

stefanozampini commented:
@tzanio Please take a look. I think I'm basically done providing support for vectors and matrices on GPU from PETSc.

camierjs commented on general/cuda.hpp:
```diff
@@ -19,6 +19,7 @@
 #define MFEM_CUDA_BLOCKS 256
 
 #ifdef MFEM_USE_CUDA
+#include <cuda.h>
```
We moved the system headers to `backends.hpp`, so that it'll be easier to switch to another backend (`hip`).
Would that work for you to include it?

stefanozampini replied:
I believe `cuda.h` belongs here since `general/cuda.hpp` uses CUDA_VERSION

camierjs replied:
I agree that there is one `include` that has sneaked in the `linalg/sparsmat.cpp` - but the goal of `backends.hpp` is to add all required system headers for the device `MFEM` has been configured for (`MFEM_USE_CUDA`, `MFEM_USE_HIP`, ...). 

When we'll want to add the `PETSc` `HIP` support, nothing will need to change - hopefully.

The `__CUDA_ARCH__` macro will be set by the `nvcc` compiler.

camierjs commented on linalg/solvers.cpp:
```diff
@@ -558,13 +558,14 @@ void CGSolver::Mult(const Vector &b, Vector &x) const
       d = r;
    }
    nom0 = nom = Dot(d, r);
+   snom = sqrt(nom);
```
I remember discussions about `sqrt` or not for this norm - we should check the reason and if we do want to do this here.

stefanozampini replied:
`CGSolver` monitoring is definitely wrong. You can check against `PetscPCG` and `HyprePCG`: now the code produces the same unpreconditioned iterates

tzanio replied:
Using the squared norm was intentional, but I agree that we should revisit it and probably change it in all the Krylov solvers.

Lets do it in another PR though. Can you revert so we can approve this one quicker?

tzanio replied:
ping: @stefanozampini

stefanozampini replied:
@tzanio Dropped the commit, it is here if you want to fix MFEM later https://github.com/mfem/mfem/tree/fix-monitoring

tzanio replied:
Thank you Stefano

camierjs commented on linalg/vector.cpp:
```diff
@@ -135,11 +135,13 @@ Vector &Vector::operator=(const Vector &v)
    // keep 'data' where it is, unless 'use_dev' is true
    if (use_dev) { Write(); }
    data.CopyFrom(v.data, v.Size());
+   v.UseDevice(vuse);
```
Why did you needed to split this `UseDev`/`CopyFrom`?

stefanozampini replied:
Because of `const` correctness, and I believe memory flags should respect that. In this case, the `MemoryFlags` of v are restored back. The current memory manager does not properly handle Synched memory (through aliases). if you remove this line ex11p will error out using a debug version of PETSc.

jandrej commented:
The changes look fine to me. The combination of MFEM + PETSc + CUDA runs flawlessly on Lassen. I don't know the details on PETSc and GPU interaction but from my perspective it looks like the majority of cases are covered.

Regarding the cmake build system support and the time frame of this PR, I'd like to get this through ASAP and then I follow up with a PR addressing cmake.

tzanio commented:
can you please merge master here and resolve the conflicts?

stefanozampini commented:
@tzanio done

camierjs commented:
Hi @stefanozampini - I've build PETSc with CUDA and MPI, to be able to compile everything, there's a function name I had to change: `MatAIJCUSPARSESetGenerateTranspose` ⇒ `MatSeqAIJCUSPARSESetGenerateTranspose`: is it the right fix to apply here?

camierjs commented:
The examples you are mentioning works well (`ex1p`, `ex9p`) with partial assembly.
`ex1p` seems to throw an error at the end of the computation when no `-pa` is given.
Both executable throw CUDA error with a `mpirun` on more than one core: is it expected within the scope of this preliminary support?

camierjs commented:
@stefanozampini , as asked by the editors, I swapped the system headers found in `linalg/sparsemat.hpp` to grab the required one from `general/backends.hpp`. The compilation is fine on my end with the `MatAIJCUSPARSESetGenerateTranspose` change too: could you make sure you are ok with these changes too?

stefanozampini commented:
What version of PETSc are you using? Can you give me the command line to reproduce the error with ex1p? better to use the dev version of PETSc, `MatAIJCUSPARSESetGenerateTranspose` is in the dev version, I'll fix it.

camierjs commented:
I used `git clone -b release https://gitlab.com/petsc/petsc.git petsc`

stefanozampini commented:
Can you use master and confirm the examples run in parallel? I haven't used `release` since GPU support has rapidly evolved in PETSc from the last release (Sept 2020).

camierjs commented:
Hum, I reverted the two names, but am hitting the same errors:
```
[~/home/petsc-cuda-support/examples/petsc]>mpirun -n 2 ./ex1p -pa -d cuda
Options used:
   --mesh ../../data/star.mesh
   --order 1
   --no-static-condensation
   --partial-assembly
   --device cuda
   --no-visualization
   --usepetsc
   --petscopts
   --no-petscmonitor
   --noforce-wrap
   --no-h2
Device configuration: cuda,cpu
Memory configuration: host-std,cuda
Number of finite element unknowns: 82561


CUDA error: (cudaGetLastError()) failed with error:
 --> no kernel image is available for execution on the device
 ... in function: void mfem::CuWrap1D(int, DBODY&&) [with int BLCK = 256; DBODY = __nv_dl_wrapper_t<__nv_dl_tag<void (mfem::ConstrainedOperator::*)(const mfem::Vector&, mfem::Vector&) const, &mfem::ConstrainedOperator::Mult, 1>, double*, const int*>&]
 ... in file: linalg/../general/forall.hpp:296
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 1 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
[0]PETSC ERROR: ------------------------------------------------------------------------
[0]PETSC ERROR: Caught signal number 15 Terminate: Some process (or the batch system) has told this process to end
[0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[0]PETSC ERROR: or see https://www.mcs.anl.gov/petsc/documentation/faq.html#valgrind
[0]PETSC ERROR: or try http://valgrind.org on GNU/linux and Apple Mac OS X to find memory corruption errors
[0]PETSC ERROR: likely location of problem given in stack below
[0]PETSC ERROR: ---------------------  Stack Frames ------------------------------------
[0]PETSC ERROR: Note: The EXACT line numbers in the stack are not available,
[0]PETSC ERROR:       INSTEAD the line number of the start of the function
[0]PETSC ERROR:       is given.
[0]PETSC ERROR: [0] __mfem_mat_shell_apply line 4937 linalg/petsc.cpp
[0]PETSC ERROR: [0] MatMult_Shell line 1062 /home/usr/src/petsc/src/mat/impls/shell/shell.c
```

stefanozampini commented:
Ok, I was able to reproduce the issue with 2 processes, but I'm not sure what is the proper fix. @tzanio @v-dobrev may know whom to address the following issue.

The device class in MFEM does not know anything about MPI, so when run with two processes, both processes call `cudaSetDevice(0)`. With your command line, PETSc device initialization is postponed up to the creation of the first device enabled object, and it tries to assign devices in round robin fashion, which means each rank gets a different device.
This issue goes away if you use  `-log_view` from PETSc, since when we use this options PETSc initializes the device during PETSc intiialize the same round robin way, with the only difference that it happens right after the instantiation of the `Device` object, so all MFEM cuda call/allocations etc will use the same device as PETSc. How are you managing multiple CUDA devices in MFEM?

v-dobrev commented:
In multi-GPU runs, the assumption in examples like `ex1p` is that the batch system assigns the visible GPUs for each MPI task (using the standard CUDA environment variable `CUDA_VISIBLE_DEVICES`). That way, for each MPI task, GPU 0 corresponds to a different physical GPU. This can also be done by the user, if the batch system does not do it. Note that in MFEM, one MPI task can work with at most 1 GPU, at least for the moment.

If each MPI task sees all GPUs indexed with their physical ids (I think this is what happens when `CUDA_VISIBLE_DEVICES` is not defined) then one needs to do the assignments in the MFEM code by setting the `dev` parameter of the `Device` constructor, e.g. in a round-robin fashion.

Does this answer your question, @stefanozampini?

stefanozampini commented:
Ok. What is the best approach then? I can initialize PETSc with whatever device MFEM is using, and this will fix the segfault, However, in this specific case ex1p will always use the same device for all ranks at least on my machine and on @camierjs. We could do round-robin assignments in ex1p directly or have the Device class handle the MPI case inside the code (when CUDA_VISIBLE_DEVICES is not set). I prefer the latter approach. Also, when running on SUMMIT with more than one GPU per resource set (say 2 MPI processes and 2 GPUs per resource set), how are you handling this case at the moment?

v-dobrev commented:
> Ok. What is the best approach then?

I don't know quite understand what the issue is. Is the problem that MFEM uses one GPU device and PETSc is using a different device? Presumably, this happens when `CUDA_VISIBLE_DEVICES` is not defined?

Generally, when `CUDA_VISIBLE_DEVICES` is not defined we need to guess what the right configuration is. For example, if I'm rank 1 of 8 (0-7) and I see 4 GPUs: do I grab GPU 1 or GPU 0? The first makes sense when ranks 0-3 are on the same node, 0, which has 4 GPUs; the second one makes sense when ranks 1,3,5,7 are on node 1 which has 4 GPUs. What if the actual job is run on 4 nodes with 2 ranks per node and each node has 4 GPUs? Maybe we can query MPI to check which other ranks are on the same node?

What does PETSc do?

> ... when running on SUMMIT with more than one GPU per resource set (say 2 MPI processes and 2 GPUs per resource set), how are you handling this case at the moment?

I have not run on Summit, so I don't know how things are configured there. If they are not setting `CUDA_VISIBLE_DEVICES`, then MFEM examples like `ex1p` will try to use the same physical GPU 0 on each node.

stefanozampini commented:
@v-dobrev PETSc currently assigns round-robin `cudaSetDevice(mpirank%devCount)`. We can also run on a specific device by using command line arguments, and these can be inserted programmatically right before PetscInitialize in MFEMInitializePETSc. The proper solution would be to extend the MFEM Device class with whatever you think is best for MFEM, then call

```
void MFEMInitializePETSc()
{
devId = Device.GetId();
PetscOptionsInsert('-cuda_device`,devId);
PetscInitialize();
}
```

This way PETSc will use the same device as the MFEM process.

v-dobrev commented:
@stefanozampini, I added the `Device::GetId()` method and some logic in `MFEMInitializePetsc` (to propagate MFEM device config to PETSc) but the latter is still incomplete -- I don't know exactly what PETSc functions to call (e.g. `PetscOptionsInsert` seems to require 4 arguments). Can you please fill in these missing parts? Thanks!

stefanozampini commented:
@v-dobrev I have finished what you started and force-pushed. Now it works properly, @camierjs Can you confirm?

camierjs commented:
The `CUDA` MPI tests are good  👍

camierjs reviewed (approved):
Great addition @stefanozampini!

Nice to see the how the `CUDA` integration was made possible through the `PetscMemory`.

As @jandrej mentioned, we might want the `cmake` and `make` build systems to be updated later in another PR: the `nvcc` `Xlinker` prefix needs to be propagated to all the `PETSc` TPLs options for example - but specifying `PETSC_OPT` and `PETSC_LIB` works just fine!

camierjs commented:
LGTM

v-dobrev commented:
@camierjs,

> As @jandrej mentioned, we might want the `cmake` and `make` build systems to be updated later in another PR: the `nvcc` `Xlinker` prefix needs to be propagated to all the `PETSc` TPLs options for example - but specifying `PETSC_OPT` and `PETSC_LIB` works just fine!

On the GNU make side there is this substitution in the `PETSC_LIB`: https://github.com/mfem/mfem/blob/9227ad8203144f41fad5c4652b4d112f6359993c/config/defaults.mk#L320

that replaces `-Wl,` with `-Xlinker=` when CUDA is enabled in MFEM. This should handle the case when PETSc is build without CUDA. When it is build with CUDA, I'm not sure what these link flags look like -- maybe they already use `-Xlinker=`?

If the GNU make and/or CMake build systems do not work properly (with some combination of CUDA/no-CUDA builds of MFEM and PETSc) without manually setting `PETSC_OPT` or `PETSC_LIB` when building MFEM, then I think we should fix that before merging the PR.

wcdawn commented:
Is there any update on this? I'd like to use MFEM+PETSc on Summit. Currently, if I use more than 1 MPI rank, (e.g. 2MPI+1GPU or 2MPI+2GPU) I get a SEGV error from PETSc. I'm testing using `ex11p` which is actually a SLEPc example but the principle is the same. I can share more about my `jsrun` or Summit submit script if that would be helpful.

Thanks!

tzanio commented:
Can you merge `master` and resolve the conflict?

stefanozampini commented:
> Can you merge `master` and resolve the conflict?

@tzanio Done. I have added a couple of fixes supporting an API change for the upcoming petsc release and added support for device_id when initializing SLEPc. This should be good to go

wcdawn commented:
I solved my problem with this. Just wanted to document here. The problem is that CUDA Aware MPI is not the default on Summit (or similar `jsrun` systems like Ascent or Sierra). The option `--smpiargs="-gpu"` must be passed to `jsrun`or else memory errors are encountered.

camierjs commented:
Hi @stefanozampini , in PETSc `main` `gitk src/mat/impls/aij/mpi/mpicusparse/mpiaijcusparse.cu` it seems that `MatAIJCUSPARSESetGenerateTranspose` has been removed just before the `3.15.0` tag.
Let me know if you confirm and if you see a way for me to test this PR again with `cmake`.

camierjs commented:
> If the GNU make and/or CMake build systems do not work properly (with some combination of CUDA/no-CUDA builds of MFEM and PETSc) without manually setting `PETSC_OPT` or `PETSC_LIB` when building MFEM, then I think we should fix that before merging the PR.

The issue I had was related to an old system blas/lapack used when compiling `PETSc`, using the `--download-f2cblaslapack` flag allowes to compile everything fine both on the `make` and `cmake` sides.

camierjs commented:
LGTM

stefanozampini commented:
> Hi @stefanozampini , in PETSc `main` `gitk src/mat/impls/aij/mpi/mpicusparse/mpiaijcusparse.cu` it seems that `MatAIJCUSPARSESetGenerateTranspose` has been removed just before the `3.15.0` tag.
> Let me know if you confirm and if you see a way for me to test this PR again with `cmake`.

@camierjs I'll take a look at it tomorrow, thanks

stefanozampini commented:
@v-dobrev @tzanio Do you think the usage of the PETSc/SLEPc interface from MFEM should be included in the CI?

stefanozampini commented:
@camierjs fixed the compilation issue

tzanio commented:
Merged in `next` for testing...

v-dobrev commented:
@stefanozampini, the internal tests show that this branch does not work with petsc v3.8.0. We see errors like this:
```console
linalg/petsc.cpp(243): error: identifier "VecCUDAGetArrayRead" is undefined
        ierr = VecCUDAGetArrayRead(x,(const PetscScalar**)&darray); PCHKERRQ(x,ierr);
               ^

linalg/petsc.cpp(245): error: identifier "VecCUDARestoreArrayRead" is undefined
        ierr = VecCUDARestoreArrayRead(x,(const PetscScalar**)&darray);
               ^

linalg/petsc.cpp(317): error: identifier "VecGetArrayWrite" is undefined
        ierr = VecGetArrayWrite(x,&v); PCHKERRQ(x,ierr);
               ^

...
```
Is this just a missing guard?

stefanozampini commented:
@v-dobrev Yep, tomorrow I will fix it. Thanks

stefanozampini commented:
@v-dobrev can you please rerun your internal tests? It should be fixed now

tzanio commented:
Merged in `next` for testing ...

v-dobrev commented:
Hi @stefanozampini,

We still get errors like these:
```console
linalg/petsc.cpp(318): error: identifier "VecGetArrayWrite" is undefined
        ierr = VecGetArrayWrite(x,&v); PCHKERRQ(x,ierr);
               ^

linalg/petsc.cpp(320): error: identifier "VecRestoreArrayWrite" is undefined
        ierr = VecRestoreArrayWrite(x,&v); PCHKERRQ(x,ierr);
               ^

...
```
I'm looking at https://gitlab.com/petsc/petsc/-/blob/v3.8/include/petscvec.h and I don't see `VecGetArrayWrite`. The first version where I see this function is v3.12.

If supporting older versions of PETSc (say, >= 3.8 and < 3.12) is not that simple, we can just change the version requirement for PETSc, as specified in `INSTALL`. In that case, we'll also update the PETSc version we are using in the nightly tests.

What do you think?

stefanozampini commented:
3.8 is from Sep 26, 2017, but it should not be hard to keep supporting it. I'll see what I can do later today. Sorry for the delay

v-dobrev commented:
> 3.8 is from Sep 26, 2017, but it should not be hard to keep supporting it. I'll see what I can do later today. Sorry for the delay

Okay, thanks!

stefanozampini commented:
@v-dobrev WIth the last commit, it should build fine in your private tests. I have tested the code in this MR from 3.8.4 to 3.14.6 and it works fine CPU side. The GPU enabled code is from 3.15 on.

tzanio commented:
Re-merged in `next` for testing ...

v-dobrev commented:
@stefanozampini, the nightly tests with PETSc 3.8 (without CUDA) look good now. Thanks!

v-dobrev commented:
LGTM

tzanio commented:
Re-merged in `next`

tzanio reviewed (approved):
Thank you @stefanozampini for this important 🔥 addition to MFEM!  

I am pretty sure a lot of our users will be excited to use it soon 😁 

_@v-dobrev, @jandrej and @camierjs -- thank you for your careful review!_
